packages/core/test/client.test.js

```diff
@@ -474,6 +499,86 @@ describe('@bugsnag/core/client', () => {
       sessionClient.notify(new Event('err', 'bad', [], { unhandled: true, severity: 'error', severityReason: { type: 'unhandledException' } }))
       sessionClient.notify(new Event('err', 'bad', [], { unhandled: true, severity: 'error', severityReason: { type: 'unhandledException' } }))
     })
+
+    it('does not start the session if onSession returns false', () => {
+      const client = new Client({ apiKey: 'API_KEY', onSession: s => false })
+      const sessionDelegate = {
+        startSession: () => {},
+        pauseSession: () => {},
+        resumeSession: () => {}
+      }
+      client._sessionDelegate = sessionDelegate
+
+      const startSpy = spyOn(sessionDelegate, 'startSession')
+
+      client.startSession()
+      expect(startSpy).toHaveBeenCalledTimes(0)
+    })
+
+    it('tolerates errors in onSession callbacks', () => {
+      const client = new Client({
+        apiKey: 'API_KEY',
+        onSession: s => {
+          throw new Error('oh no')
+        }
+      })
+      const sessionDelegate = {
+        startSession: () => {},
+        pauseSession: () => {},
+        resumeSession: () => {}
+      }
+      client._sessionDelegate = sessionDelegate
+
+      const startSpy = spyOn(sessionDelegate, 'startSession')
+
+      client.startSession()
+      expect(startSpy).toHaveBeenCalledTimes(1)
+    })
+  })
+
+  describe('callbacks', () => {
+    it('supports adding and removing onError/onSession/onBreadcrumb callbacks', (done) => {
+      const c = new Client({ apiKey: 'API_KEY' })
+      c._setDelivery(client => ({ sendEvent: (p, cb) => cb(null), sendSession: (s, cb) => cb(null) }))
+      c._logger = console
+      const sessionDelegate = {
+        startSession: () => {},
+        pauseSession: () => {},
+        resumeSession: () => {}
+      }
+      c._sessionDelegate = sessionDelegate
+      const eSpy = spyOn({ fn: () => {} }, 'fn')
+      const bSpy = spyOn({ fn: () => {} }, 'fn')
+      const sSpy = spyOn({ fn: () => {} }, 'fn')
+
+      c.addOnError(eSpy)
+      c.addOnSession(sSpy)
+      c.addOnBreadcrumb(bSpy)
+
+      expect(c._cbs.e.length).toBe(1)
+      expect(c._cbs.s.length).toBe(1)
+      expect(c._cbs.b.length).toBe(1)
+
+      c.startSession()
+      expect(sSpy).toHaveBeenCalledTimes(1)
+      c.notify(new Error(), () => {}, () => {
+        expect(bSpy).toHaveBeenCalledTimes(1)
+        expect(bSpy).toHaveBeenCalledTimes(1)
+
+        c.removeOnError(eSpy)
+        c.removeOnSession(sSpy)
+        c.removeOnBreadcrumb(bSpy)
+
+        c.startSession()
+        expect(sSpy).toHaveBeenCalledTimes(1)
+        c.notify(new Error(), () => {}, () => {
+          expect(bSpy).toHaveBeenCalledTimes(1)
+          expect(bSpy).toHaveBeenCalledTimes(1)
```

Comment:
duplicate assertion

Reply:
Good spot – this was actually meant to make an assertion about `eSpy`.